File: 838.pushDominoes/pushDominoes.java
class Solution {
    public String pushDominoes(String dominoes) {
        String d = 'L' + dominoes + 'R';
        int N = d.length();
        int left = 0, right = 1;
        StringBuilder sb = new StringBuilder();
        while (right < N) {
            if (d.charAt(right) == '.') {
                right++;
                continue;
            }
            int dotNum = right - left - 1;
            if (d.charAt(right) == 'L') {
                if (d.charAt(left) == 'L') { // L...L
                    if (left != 0) sb.append('L');
                    for (int i = 0; i < dotNum; i++) sb.append('L');
                } else if (d.charAt(left) == 'R') { // R...L
                    sb.append('R');
                    for (int i = 0; i < dotNum / 2; i++) sb.append('R');
                    if (dotNum % 2 != 0) sb.append('.');
                    for (int i = 0; i < dotNum / 2; i++) sb.append('L');
                }
            } else if (d.charAt(right) == 'R') {
                if (d.charAt(left) == 'L') { // L...R
                    if (left != 0) sb.append('L');
                    for (int i = 0; i < dotNum; i++) sb.append('.');
                } else if (d.charAt(right) == 'R') { // R...R
                    for (int i = 0; i < dotNum + 1; i++) sb.append('R');
                }
            }
            left = right;
            right = left + 1;
        }
        return sb.toString();
    }
}

File: 838.pushDominoes/pushDominoes.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    string pushDominoes(string dominoes) {
        dominoes = "L" + dominoes + "R";
        int len = dominoes.length();
        int left = 0, right = 1;
        string res;
        while (right < len) {
            if (dominoes[right] == '.') {
                right++;
                continue;
            }
            if (left > 0)
                res += dominoes[left];
            int mid_len = right - left - 1;
            if (dominoes[right] == 'L') {
                if (dominoes[left] == 'L') {
                    res += string(mid_len, 'L');
                } else if (dominoes[left] == 'R') {
                    res += string(mid_len / 2, 'R') + string(mid_len % 2, '.') + string(mid_len / 2, 'L');
                }
            } else if (dominoes[right] == 'R') {
                if (dominoes[left] == 'L') {
                    res += string(mid_len, '.');
                } else if (dominoes[left] == 'R') {
                    res += string(mid_len, 'R');
                }
            }
            left = right;
            right++;
        }
        return res;
    }
};

int main() {
    Solution s;
    string dominoes = ".L.R...LR..L..";
    cout << s.pushDominoes(dominoes) << endl;
    return 0;
}
